### PR TITLE
Fix intoto index keys

### LINKED ISSUE #876: Intoto Indexing Bug: cannot find entry by payloadHash
**Description**

Currently at head, we cannot use the search index to find entries by payloadHash, which should be indexed: 
https://github.com/sigstore/rekor/blob/744339f1be49454bb3009bd0d31327f8d53646cd/pkg/types/intoto/v0.0.1/entry.go#L77-L78

**Version**

I spun up a local rekor instance and uploaded an intoto attestation. I logged what keys were added. Check out the payloadHash in all cases:

The entry:
```
Body: {
  "IntotoObj": {
    "content": {
      "hash": {
        "algorithm": "sha256",
        "value": "51a0a4f22ff245835d264cd41c6969a58fb3e37884c7aa2cace767714beb2a22"
      },
      "payloadHash": {
        "algorithm": "sha256",
        "value": "d941d60e3f06985f970096cc0ff5ae2990ae64eab57230e5f9c9f4cf89d4e5de"
      }
    },
    "publicKey": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURRRENDQXNXZ0F3SUJBZ0lVQU9ySkxadHRpTXFZUm1nWHdRczVva3FHczlNd0NnWUlLb1pJemowRUF3TXcKS2pFVk1CTUdBMVVFQ2hNTWMybG5jM1J2Y21VdVpHVjJNUkV3RHdZRFZRUURFd2h6YVdkemRHOXlaVEFlRncweQpNakEyTVRZeE16STJNREZhRncweU1qQTJNVFl4TXpNMk1EQmFNQUF3V1RBVEJnY3Foa2pPUFFJQkJnZ3Foa2pPClBRTUJCd05DQUFUUlQ4WlFMZnl4RThNZE5DcmhsZWNnQmFyWUs3WjJvb0c1SllCaElwV3IwaTRFSVEzSGg2bSsKYThrcnZVWFFnSmlGclNRdFl6V0EyWlk3WEU2Zi9HNFhvNElCOFRDQ0FlMHdEZ1lEVlIwUEFRSC9CQVFEQWdlQQpNQk1HQTFVZEpRUU1NQW9HQ0NzR0FRVUZCd01ETUF3R0ExVWRFd0VCL3dRQ01BQXdIUVlEVlIwT0JCWUVGSU91ClZEK01RcDJiRDk1c3UwMzN1MTU5bDliT01COEdBMVVkSXdRWU1CYUFGRmpBSGwrUlJhVm1xWHJNa0tHVEl0QXEKeGNYNk1IMEdBMVVkRVFFQi93UnpNSEdHYjJoMGRIQnpPaTh2WjJsMGFIVmlMbU52YlM5emJITmhMV1p5WVcxbApkMjl5YXk5emJITmhMV2RwZEdoMVlpMW5aVzVsY21GMGIzSXZMbWRwZEdoMVlpOTNiM0pyWm14dmQzTXZZblZwCmJHUmxjbDluYjE5emJITmhNeTU1Yld4QWNtVm1jeTkwWVdkekwzWXhMakF1TURBZkJnb3JCZ0VFQVlPL01BRUMKQkJGM2IzSnJabXh2ZDE5a2FYTndZWFJqYURBZEJnb3JCZ0VFQVlPL01BRUVCQTlIYnlCVFRGTkJJRkpsYkdWaApjMlV3T1FZS0t3WUJCQUdEdnpBQkFRUXJhSFIwY0hNNkx5OTBiMnRsYmk1aFkzUnBiMjV6TG1kcGRHaDFZblZ6ClpYSmpiMjUwWlc1MExtTnZiVEEyQmdvckJnRUVBWU8vTUFFREJDZ3laRFEzTldRM05HRTVNMlV6TnpWaVptWXoKTkRJd1pqTXdPVE13TXpNMFlUaGpPRGhpTjJVeU1CMEdDaXNHQVFRQmc3OHdBUVlFRDNKbFpuTXZhR1ZoWkhNdgpiV0ZwYmpBbkJnb3JCZ0VFQVlPL01BRUZCQmxoYzNKaFlTOXpiSE5oTFc5dUxXZHBkR2gxWWkxMFpYTjBNQW9HCkNDcUdTTTQ5QkFNREEya0FNR1lDTVFEMGVpU09FZWVoenVXWlhwWHVYMlJvVEFEY3UrUG9NSEVpb0NYalQwUjkKeE03VzBtQkFZZmVyUmk0YmxmQkxXVVlDTVFDdzhiWjBsaTV2eE9yd25rVFIxU2JsU1lOWk91b2Q1SC8vK09vZwo2WHdXeUp4cnNzcm8rT0RRcHZuYzB0VXB2Szg9Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K"
  }
}
```

The indexed keys:
```
 sha256:fe24a8c9e6a0eab15c97c70c3d1e56f2dd36b36c231c544c75374a7c96c4d8ac (payloadHash), 

sha256:51a0a4f22ff245835d264cd41c6969a58fb3e37884c7aa2cace767714beb2a22, (envelope digest)

sha256:4bdb495d05cbb9eb0722ec4fcfef8a02d5ca7a2fca0b0fed718447f89d39a589, (subject digest)

sha1:2d475d74a93e375bff3420f30930334a8c88b7e2 (materials)
```

Note that the payload hashes do NOT match

1. Calculated for the entry [here](https://github.com/sigstore/rekor/blob/4f0b750ea3cee548684227570190250fcd136f02/pkg/types/intoto/v0.0.1/entry.go#L183-L189) via [AttestationKeyValue](https://github.com/sigstore/rekor/blob/4f0b750ea3cee548684227570190250fcd136f02/pkg/types/intoto/v0.0.1/entry.go#L248-L249)
2. Calculated for the index [here](https://github.com/sigstore/rekor/blob/4f0b750ea3cee548684227570190250fcd136f02/pkg/types/intoto/v0.0.1/entry.go#L76-L77)

### LINKED ISSUE #890: Index intoto attestations by email
I've got a Rekor entry e305d5055749770fbfb067466f65d4879f77618f1263e29d78a350464af144db that contains an intoto attestation, which is signed by a Fulcio cert for 7577580991-compute@developer.gserviceaccount.com.

I can get the entry:

```
$ rekor-cli get --uuid e305d5055749770fbfb067466f65d4879f77618f1263e29d78a350464af144db
LogID: c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d
Attestation: {"_type":"intoto","predicateType":"tlogistry-fetched","subject":[{"name":"ghcr.io/distroless/static:latest-20220601","digest":{"sha256":"5663c7476101e26401b52a92d9b7adce6acd6036ebc330f0e60c828585e5e9c1"}}],"predicate":{"digest":"sha256:5b0dd45ad9a90361d4224850d73c8e0b9d87454772c13605ff6548f8ef88f4fa","tag":"ghcr.io/distroless/static:latest-20220601"}}
...
```

But I can't search by email and find the entry:

```
$ rekor-cli search --email 7577580991-compute@developer.gserviceaccount.com
no matching entries found
```

In Slack @bobcallaway seemed to indicate this was an oversight, and not intentional.

Filing a bug to track this. If someone can point me to the code where this happens I can try to make the change.

### PR DESCRIPTION
**Pull request by bobcallaway**
Ensure we include all appropriate index keys including:
- entire DSSE envelope SHA256 digest
- envelope (base64 decoded) SHA256 digest
- entire X509 signing certificate
- any relevant keys extracted from X509 signing certificate

Fixes: #890 
Fixes: #876

Signed-off-by: Bob Callaway <bcallaway@google.com>

### CODE REVIEW

**Review by dlorenc — DISMISSED** [review]

File: pkg/types/intoto/v0.0.1/entry_test.go [hunk 1]
```diff
@@ -227,13 +249,41 @@ func TestV001Entry_Unmarshal(t *testing.T) {
...
+				canonicalEntry, err := types.NewEntry(pe)
+				if err != nil {
+					t.Errorf("unexpected err from type-specific unmarshalling for '%v': %v", tt.name, err)
+				}
+				canonicalV001 := canonicalEntry.(*V001Entry)
+				fmt.Printf("%v", canonicalV001.IntotoObj.Content)
+				if *canonicalV001.IntotoObj.Content.Hash.Value != *tt.it.Content.Hash.Value {
+					t.Errorf("envelope hashes do not match post canonicalization: %v %v", *canonicalV001.IntotoObj.Content.Hash.Value, *tt.it.Content.Hash.Value)
+				}
+				if canonicalV001.AttestationKey() != "" && *canonicalV001.IntotoObj.Content.PayloadHash.Value != payloadHash {
```
Comment (asraa) [inline pkg/types/intoto/v0.0.1/entry_test.go:285]: NICE

File: pkg/types/intoto/v0.0.1/entry.go [hunk 2]
```diff
@@ -73,8 +73,24 @@ func NewEntry() types.EntryImpl {
...
+		result = append(result, "sha256:"+strings.ToLower(hex.EncodeToString(keyHash[:])))
+
+		pub, err := x509.NewPublicKey(bytes.NewReader(key))
+		if err != nil {
+			return result, err
+		}
+		result = append(result, pub.EmailAddresses()...)
+	}
+
+	payloadBytes, err := v.env.DecodeB64Payload()
```
Comment (asraa) [inline pkg/types/intoto/v0.0.1/entry.go]: Can this be used https://github.com/sigstore/rekor/blob/ceb43c5f2edf5c6347994e017cd533f5ac0cbd3b/pkg/types/intoto/v0.0.1/entry.go#L263 for consistency?

**Review by asraa — DISMISSED** [review]

File: pkg/types/intoto/v0.0.1/entry.go (hunk 2 above)
Comment (bobcallaway) [reply pkg/types/intoto/v0.0.1/entry.go]: sure, can do that

**Review by asraa — DISMISSED** [review]

File: pkg/types/intoto/v0.0.1/entry.go [hunk 3]
```diff
@@ -73,20 +73,42 @@ func NewEntry() types.EntryImpl {
...
+	// add digest over entire DSSE envelope
+	if v.IntotoObj.Content != nil && v.IntotoObj.Content.Hash != nil {
+		hashkey := strings.ToLower(fmt.Sprintf("%s:%s", swag.StringValue(v.IntotoObj.Content.Hash.Algorithm), swag.StringValue(v.IntotoObj.Content.Hash.Value)))
+		result = append(result, hashkey)
+	} else {
+		log.Logger.Error("could not find content digest to include in index keys")
+	}
+
+	// add digest over public key
+	if v.IntotoObj.PublicKey != nil && v.keyObj != nil {
```
Comment (asraa) [inline pkg/types/intoto/v0.0.1/entry.go]: Resolve this:
Had to check this for myself: `keyObj` is populated in https://github.com/sigstore/rekor/blob/a166f1de5dbc1090bd27443db1f56be39d02ede0/pkg/types/intoto/v0.0.1/entry.go#L154

before `IndexKeys()` is called.

File: pkg/types/intoto/v0.0.1/entry.go [hunk 4]
```diff
@@ -73,20 +73,42 @@ func NewEntry() types.EntryImpl {
...
+		key := *v.IntotoObj.PublicKey
+		keyHash := sha256.Sum256(key)
+		result = append(result, "sha256:"+strings.ToLower(hex.EncodeToString(keyHash[:])))
```
Comment (asraa) [inline pkg/types/intoto/v0.0.1/entry.go]: style nit: use `fmt.Sprintf` like above and below? Don't care too much though. Clear enough

File: pkg/types/intoto/v0.0.1/entry.go [hunk 5]
```diff
@@ -73,20 +73,42 @@ func NewEntry() types.EntryImpl {
...
+	if v.IntotoObj.Content != nil && v.IntotoObj.Content.Hash != nil {
+		hashkey := strings.ToLower(fmt.Sprintf("%s:%s", swag.StringValue(v.IntotoObj.Content.Hash.Algorithm), swag.StringValue(v.IntotoObj.Content.Hash.Value)))
+		result = append(result, hashkey)
+	} else {
+		log.Logger.Error("could not find content digest to include in index keys")
+	}
+
+	// add digest over public key
+	if v.IntotoObj.PublicKey != nil && v.keyObj != nil {
+		key := *v.IntotoObj.PublicKey
```
Comment (asraa) [inline pkg/types/intoto/v0.0.1/entry.go]: Use `CanonicalValue()`?

To match https://github.com/sigstore/rekor/blob/408b752c60f67e0efca3e8965360dfe28f957cdf/pkg/api/index.go#L63-L68

**Review by dlorenc — APPROVED** [review]

**Comment by dlorenc** [comment]
cc @cpanato we should cut a release after this goes in I think

**Comment by cpanato** [comment]
Ok @dlorenc will prepare the changelog and check if need other updates

**Pull request merged**